erpnext/controllers/accounts_controller.py

```diff
@@ -1087,9 +1217,15 @@ def update_against_document_in_jv(self):
 			reconcile_against_document(lst)
 
 	def on_cancel(self):
-		from erpnext.accounts.utils import unlink_ref_doc_from_payment_entries
+		from erpnext.accounts.utils import (
+			cancel_exchange_gain_loss_journal,
+			unlink_ref_doc_from_payment_entries,
+		)
+
+		if self.doctype in ["Sales Invoice", "Purchase Invoice", "Payment Entry"]:
```

Comment:
How are payments via Journal Entry handled? Or is this only limited to payment entry for now?

Reply:
This logic has been refactored. New logic handles Journal Entry as well.